### DISCUSSION TITLE
How to invert a matrix?

### DISCUSSION THREAD

**Discussion by ghost** [modula3/cm3#1072] [General]
Is there any built-in functions to do so?

**Comment by jpgpng** [comment]
It seems he wrote the matrix library. Let's call him @mikanystrom

**Reply by ghost** [reply]
There are two matrix libraries. One is from caltech-other (written by the one you mentioned) and one from m3-libs linearalgebra. Can't find the inverse matrix function on either of them.

**Comment by mikanystrom** [comment]
Use the LU decomposition library from caltech-other, for example.

https://www2.cpfs.mpg.de/~kasinath/pqm1/lecturenotes/lecture1/NumericalRecipesinF77.pdf

read Section 2.3.

**Reply by ghost** [reply]
Is there any ready to use functions? It's straight forward on Python, just call the only one function and it's done. What you can't make things simpler and just work?

**Comment by thielema** [comment]
On Sat, 10 Sep 2022, pwd96m4a wrote:

> Is there any ready to use functions? It's straight forward on Python, 
> just call the only one function and it's done. What you can't make 
> things simpler and just work?

There is the 'arithmetic' package and a binding to LAPACK.

**Comment by mikanystrom** [comment]
Usually you don't actually want to invert a matrix, except on an exam.

If you're trying to solve a system of equations, there are better
algorithms. Gauss-Jordan for example. LU is even better.  Multiplying by
the inverse is far inferior to either both in terms of speed and precision
(partly because the determinant is a high power for large matrices, 1/D can
overflow or underflow quite easily).

If the matrix has special properties, such that it actually makes sense to
invert it, usually you want to use even more specialized algorithms.

In any case, the section of Numerical Recipes i referenced shows how to use
LU to invert a matrix if for some strange reason that's what you really
want. It's a couple of lines of code.

    Mika

**Comment by mikanystrom** [comment]
hah.. yes there is a Matrix library in there that I wrote.

There are as many ways to invert a matrix as there are to skin a cat...
each having different purposes.

https://github.com/modula3/cm3/blob/master/caltech-other/matrix/src/LU.m3

https://github.com/modula3/cm3/blob/master/caltech-other/matrix/src/LU2.mg

There is a comment on line 14.  It's a reference, really.  Obviously this
code didn't actually come from a Fortran book.  But it works the same.

I think there's a Fortran implementation in here, too, compiled to assembly
language on i386, I think.  Hmm.. I guess it doesn't use the Fortran code
on amd64.

There's also SVD, and if you look around, I think there's probably Cholesky
decomposition as well.

Now, what are you trying to do with matrix inversion?  Usually that's just
part of some larger algorithm.  There's a Newton root finder in the same
directory.  The secant method is implemented somewhere else, and some fancy
multi-dimensional optimizers...

     Mika
